File: testers/duckdb/scala/src/scala/testdb/CompositeKeyTest.scala
package testdb

import org.junit.Assert._
import org.junit.Test
import testdb.departments._
import testdb.employees._

import java.time.LocalDate

/** Tests for composite primary keys in DuckDB. Tests the departments (2-column String,String) and employees (2-column Integer,String) tables with composite PKs.
  */
class CompositeKeyTest {
  private val departmentsRepo = DepartmentsRepoImpl()
  private val employeesRepo = EmployeesRepoImpl()

  // ==================== Departments (String, String) Composite Key ====================

  @Test
  def testDepartmentsInsert(): Unit = withConnection {
    val unique = System.nanoTime().toString.takeRight(6)
    val dept = DepartmentsRow(s"IT$unique", "US-WEST", "Information Technology", Some(BigDecimal("1000000")))

    val inserted = departmentsRepo.insert(dept)

    assertNotNull(inserted)
    assertEquals(s"IT$unique", inserted.deptCode)
    assertEquals("US-WEST", inserted.deptRegion)
    assertEquals("Information Technology", inserted.deptName)
    assertTrue(inserted.budget.isDefined)
    assertEquals(0, inserted.budget.get.compareTo(BigDecimal("1000000")))
  }

  @Test
  def testDepartmentsSelectByCompositeId(): Unit = withConnection {
    val dept = DepartmentsRow("HR", "EU-EAST", "Human Resources", None)
    val _ = departmentsRepo.insert(dept)

    val id = DepartmentsId("HR", "EU-EAST")
    val found = departmentsRepo.selectById(id)

    assertTrue(found.isDefined)
    assertEquals("HR", found.get.deptCode)
    assertEquals("EU-EAST", found.get.deptRegion)
  }

  @Test
  def testDepartmentsCompositeIdFromRow(): Unit = withConnection {
    val dept = DepartmentsRow("SALES", "APAC", "Sales APAC", Some(BigDecimal("500000")))
    val inserted = departmentsRepo.insert(dept)

    val compositeId = inserted.compositeId
    assertEquals("SALES", compositeId.deptCode)
    assertEquals("APAC", compositeId.deptRegion)

    val found = departmentsRepo.selectById(compositeId)
    assertTrue(found.isDefined)
    assertEquals(inserted, found.get)
  }

  @Test
  def testDepartmentsUpdate(): Unit = withConnection {
    val dept = DepartmentsRow("FINANCE", "US-CENTRAL", "Finance", Some(BigDecimal("800000")))
    val inserted = departmentsRepo.insert(dept)

    val updated = inserted.copy(budget = Some(BigDecimal("900000")))
    val wasUpdated = departmentsRepo.update(updated)
    assertTrue(wasUpdated)

    val found = departmentsRepo.selectById(inserted.compositeId).get
    assertEquals(Some(BigDecimal("900000")), found.budget)
  }

  @Test
  def testDepartmentsDelete(): Unit = withConnection {
    val dept = DepartmentsRow("TEMP", "TEMP-REGION", "Temporary", None)
    val inserted = departmentsRepo.insert(dept)

    val deleted = departmentsRepo.deleteById(inserted.compositeId)
    assertTrue(deleted)

    val found = departmentsRepo.selectById(inserted.compositeId)
    assertFalse(found.isDefined)
  }

  @Test
  def testDepartmentsMultipleSameCode(): Unit = withConnection {
    val dept1 = DepartmentsRow("ENG", "US", "Engineering US", None)
    val dept2 = DepartmentsRow("ENG", "EU", "Engineering EU", None)
    val dept3 = DepartmentsRow("ENG", "APAC", "Engineering APAC", None)

    val _ = departmentsRepo.insert(dept1)
    val _ = departmentsRepo.insert(dept2)
    val _ = departmentsRepo.insert(dept3)

    assertTrue(departmentsRepo.selectById(DepartmentsId("ENG", "US")).isDefined)
    assertTrue(departmentsRepo.selectById(DepartmentsId("ENG", "EU")).isDefined)
    assertTrue(departmentsRepo.selectById(DepartmentsId("ENG", "APAC")).isDefined)
  }

  // ==================== Employees (Integer, String) Composite Key ====================

  @Test
  def testEmployeesInsert(): Unit = withConnection {
    val unique = System.nanoTime().toString.takeRight(6)
    val empNum = (System.nanoTime() % 1000000).toInt
    val dept = DepartmentsRow(s"DEV$unique", "US", "Development", None)
    val _ = departmentsRepo.insert(dept)

    val emp = EmployeesRow(
      empNum,
      "A",
      s"DEV$unique",
      "US",
      "Alice Johnson",
      Some(BigDecimal("95000")),
      LocalDate.of(2025, 1, 15)
    )

    val inserted = employeesRepo.insert(emp)

    assertNotNull(inserted)
    assertEquals(empNum, inserted.empNumber)
    assertEquals("A", inserted.empSuffix)
    assertEquals("Alice Johnson", inserted.empName)
  }

  @Test
  def testEmployeesSelectByCompositeId(): Unit = withConnection {
    val _ = departmentsRepo.insert(DepartmentsRow("QA", "EU", "Quality Assurance", None))

    val emp = EmployeesRow(
      2001,
      "B",
      "QA",
      "EU",
      "Bob Smith",
      Some(BigDecimal("75000")),
      LocalDate.of(2024, 6, 1)
    )
    val _ = employeesRepo.insert(emp)

    val id = EmployeesId(2001, "B")
    val found = employeesRepo.selectById(id)

    assertTrue(found.isDefined)
    assertEquals(2001, found.get.empNumber)
    assertEquals("B", found.get.empSuffix)
    assertEquals("Bob Smith", found.get.empName)
  }

  @Test
  def testEmployeesCompositeIdFromRow(): Unit = withConnection {
    val _ = departmentsRepo.insert(DepartmentsRow("SUPPORT", "US", "Support", None))

    val emp = EmployeesRow(
      3001,
      "C",
      "SUPPORT",
      "US",
      "Carol White",
      None,
      LocalDate.now()
    )
    val inserted = employeesRepo.insert(emp)

    val compositeId = inserted.compositeId
    assertEquals(3001, compositeId.empNumber)
    assertEquals("C", compositeId.empSuffix)

    val found = employeesRepo.selectById(compositeId)
    assertTrue(found.isDefined)
  }

  @Test
  def testEmployeesUpdate(): Unit = withConnection {
    val _ = departmentsRepo.insert(DepartmentsRow("MGMT", "US", "Management", None))

    val emp = EmployeesRow(
      4001,
      "D",
      "MGMT",
      "US",
      "David Brown",
      Some(BigDecimal("120000")),
      LocalDate.now()
    )
    val inserted = employeesRepo.insert(emp)

    val updated = inserted.copy(salary = Some(BigDecimal("130000")))
    val _ = employeesRepo.update(updated)

    val found = employeesRepo.selectById(inserted.compositeId).get
    assertEquals(Some(BigDecimal("130000")), found.salary)
  }

  @Test
  def testEmployeesDelete(): Unit = withConnection {
    val _ = departmentsRepo.insert(DepartmentsRow("TEMP", "TEMP", "Temp Dept", None))

    val emp = EmployeesRow(5001, "X", "TEMP", "TEMP", "To Be Deleted", None, LocalDate.now())
    val inserted = employeesRepo.insert(emp)

    val deleted = employeesRepo.deleteById(inserted.compositeId)
    assertTrue(deleted)

    val found = employeesRepo.selectById(inserted.compositeId)
    assertFalse(found.isDefined)
  }

  @Test
  def testEmployeesMultipleSameNumber(): Unit = withConnection {
    val _ = departmentsRepo.insert(DepartmentsRow("SHARED", "US", "Shared Dept", None))

    val emp1 = EmployeesRow(9999, "A", "SHARED", "US", "Employee A", None, LocalDate.now())
    val emp2 = EmployeesRow(9999, "B", "SHARED", "US", "Employee B", None, LocalDate.now())
    val emp3 = EmployeesRow(9999, "C", "SHARED", "US", "Employee C", None, LocalDate.now())

    val _ = employeesRepo.insert(emp1)
    val _ = employeesRepo.insert(emp2)
    val _ = employeesRepo.insert(emp3)

    assertTrue(employeesRepo.selectById(EmployeesId(9999, "A")).isDefined)
    assertTrue(employeesRepo.selectById(EmployeesId(9999, "B")).isDefined)
    assertTrue(employeesRepo.selectById(EmployeesId(9999, "C")).isDefined)
  }

  // ==================== Composite FK Relationship ====================

  @Test
  def testEmployeeDepartmentForeignKey(): Unit = withConnection {
    val dept = DepartmentsRow("FK_TEST", "FK_REGION", "FK Test Dept", Some(BigDecimal("500000")))
    val _ = departmentsRepo.insert(dept)

    val emp = EmployeesRow(
      6001,
      "F",
      "FK_TEST",
      "FK_REGION",
      "FK Employee",
      Some(BigDecimal("60000")),
      LocalDate.now()
    )
    val insertedEmp = employeesRepo.insert(emp)

    assertEquals(dept.deptCode, insertedEmp.deptCode)
    assertEquals(dept.deptRegion, insertedEmp.deptRegion)

    val foundDept = departmentsRepo.selectById(DepartmentsId(insertedEmp.deptCode, insertedEmp.deptRegion))
    assertTrue(foundDept.isDefined)
    assertEquals("FK Test Dept", foundDept.get.deptName)
  }
}

File: testers/sqlite/java/src/java/testdb/CompositeKeyTest.java
package testdb;

import static org.junit.Assert.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Optional;
import org.junit.Test;
import testdb.departments.*;
import testdb.employees.*;

/**
 * Composite primary keys: departments (VARCHAR, VARCHAR) and employees (INTEGER, VARCHAR), plus the
 * composite foreign key from employees → departments.
 *
 * <p>Schema seeds two departments (IT/US-WEST, HR/US-EAST) and two employees (1001/A, 1002/B); each
 * test runs inside a per-test rollback so it sees those rows but its own inserts vanish on exit.
 */
public class CompositeKeyTest {
  private final DepartmentsRepoImpl departmentsRepo = new DepartmentsRepoImpl();
  private final EmployeesRepoImpl employeesRepo = new EmployeesRepoImpl();

  // ==================== Departments (String, String) Composite Key ====================

  @Test
  public void testDepartmentsInsert() {
    SqliteTestHelper.run(
        c -> {
          var dept =
              new DepartmentsRow(
                  "ENG", "EU-NORTH", "Engineering Europe", Optional.of(new BigDecimal("750000")));
          var inserted = departmentsRepo.insert(dept, c);

          assertNotNull(inserted);
          assertEquals("ENG", inserted.deptCode());
          assertEquals("EU-NORTH", inserted.deptRegion());
          assertEquals("Engineering Europe", inserted.deptName());
          assertEquals(0, new BigDecimal("750000").compareTo(inserted.budget().orElseThrow()));
        });
  }

  @Test
  public void testDepartmentsSelectByCompositeId() {
    SqliteTestHelper.run(
        c -> {
          var id = new DepartmentsId("IT", "US-WEST"); // from seed
          var found = departmentsRepo.selectById(id, c);

          assertTrue(found.isPresent());
          assertEquals("IT", found.get().deptCode());
          assertEquals("US-WEST", found.get().deptRegion());
          assertEquals("Information Technology", found.get().deptName());
        });
  }

  @Test
  public void testDepartmentsCompositeIdFromRow() {
    SqliteTestHelper.run(
        c -> {
          var dept =
              new DepartmentsRow(
                  "SALES", "APAC", "Sales APAC", Optional.of(new BigDecimal("500000")));
          var inserted = departmentsRepo.insert(dept, c);

          var compositeId = inserted.compositeId();
          assertEquals("SALES", compositeId.deptCode());
          assertEquals("APAC", compositeId.deptRegion());

          var found = departmentsRepo.selectById(compositeId, c);
          assertTrue(found.isPresent());
          assertEquals(inserted, found.get());
        });
  }

  @Test
  public void testDepartmentsUpdate() {
    SqliteTestHelper.run(
        c -> {
          var dept =
              new DepartmentsRow(
                  "FINANCE", "US-CENTRAL", "Finance", Optional.of(new BigDecimal("800000")));
          var inserted = departmentsRepo.insert(dept, c);

          var updated = inserted.withBudget(Optional.of(new BigDecimal("900000")));
          boolean wasUpdated = departmentsRepo.update(updated, c);
          assertTrue(wasUpdated);

          var found = departmentsRepo.selectById(inserted.compositeId(), c).orElseThrow();
          assertEquals(0, new BigDecimal("900000").compareTo(found.budget().orElseThrow()));
        });
  }

  @Test
  public void testDepartmentsDelete() {
    SqliteTestHelper.run(
        c -> {
          var dept = new DepartmentsRow("TEMP", "TEMP-REGION", "Temporary", Optional.empty());
          var inserted = departmentsRepo.insert(dept, c);

          boolean deleted = departmentsRepo.deleteById(inserted.compositeId(), c);
          assertTrue(deleted);

          var found = departmentsRepo.selectById(inserted.compositeId(), c);
          assertFalse(found.isPresent());
        });
  }

  @Test
  public void testDepartmentsMultipleSameCode() {
    SqliteTestHelper.run(
        c -> {
          var d1 = new DepartmentsRow("ENG2", "US", "Engineering US", Optional.empty());
          var d2 = new DepartmentsRow("ENG2", "EU", "Engineering EU", Optional.empty());
          var d3 = new DepartmentsRow("ENG2", "APAC", "Engineering APAC", Optional.empty());

          departmentsRepo.insert(d1, c);
          departmentsRepo.insert(d2, c);
          departmentsRepo.insert(d3, c);

          assertTrue(departmentsRepo.selectById(new DepartmentsId("ENG2", "US"), c).isPresent());
          assertTrue(departmentsRepo.selectById(new DepartmentsId("ENG2", "EU"), c).isPresent());
          assertTrue(departmentsRepo.selectById(new DepartmentsId("ENG2", "APAC"), c).isPresent());
        });
  }

  // ==================== Employees (Long, String) Composite Key ====================

  @Test
  public void testEmployeesInsert() {
    SqliteTestHelper.run(
        c -> {
          // Schema already has IT/US-WEST as a department, reuse it for FK
          var emp =
              new EmployeesRow(
                  2001L,
                  "A",
                  "IT",
                  "US-WEST",
                  "Alice Johnson",
                  Optional.of(new BigDecimal("95000")),
                  LocalDate.of(2025, 1, 15));
          var inserted = employeesRepo.insert(emp, c);

          assertNotNull(inserted);
          assertEquals(Long.valueOf(2001L), inserted.empNumber());
          assertEquals("A", inserted.empSuffix());
          assertEquals("Alice Johnson", inserted.empName());
        });
  }

  @Test
  public void testEmployeesSelectByCompositeId() {
    SqliteTestHelper.run(
        c -> {
          var id = new EmployeesId(1001L, "A"); // from seed
          var found = employeesRepo.selectById(id, c);

          assertTrue(found.isPresent());
          assertEquals(Long.valueOf(1001L), found.get().empNumber());
          assertEquals("A", found.get().empSuffix());
          assertEquals("Alice Johnson", found.get().empName());
        });
  }

  @Test
  public void testEmployeesCompositeIdFromRow() {
    SqliteTestHelper.run(
        c -> {
          var emp =
              new EmployeesRow(
                  3001L,
                  "C",
                  "HR",
                  "US-EAST",
                  "Carol White",
                  Optional.empty(),
                  LocalDate.of(2025, 3, 1));
          var inserted = employeesRepo.insert(emp, c);

          var compositeId = inserted.compositeId();
          assertEquals(Long.valueOf(3001L), compositeId.empNumber());
          assertEquals("C", compositeId.empSuffix());

          var found = employeesRepo.selectById(compositeId, c);
          assertTrue(found.isPresent());
        });
  }

  @Test
  public void testEmployeesUpdate() {
    SqliteTestHelper.run(
        c -> {
          var emp =
              new EmployeesRow(
                  4001L,
                  "D",
                  "IT",
                  "US-WEST",
                  "David Brown",
                  Optional.of(new BigDecimal("120000")),
                  LocalDate.of(2025, 4, 1));
          var inserted = employeesRepo.insert(emp, c);

          var updated = inserted.withSalary(Optional.of(new BigDecimal("130000")));
          employeesRepo.update(updated, c);

          var found = employeesRepo.selectById(inserted.compositeId(), c).orElseThrow();
          assertEquals(0, new BigDecimal("130000").compareTo(found.salary().orElseThrow()));
        });
  }

  @Test
  public void testEmployeesDelete() {
    SqliteTestHelper.run(
        c -> {
          var emp =
              new EmployeesRow(
                  5001L,
                  "X",
                  "IT",
                  "US-WEST",
                  "To Be Deleted",
                  Optional.empty(),
                  LocalDate.of(2025, 5, 1));
          var inserted = employeesRepo.insert(emp, c);

          boolean deleted = employeesRepo.deleteById(inserted.compositeId(), c);
          assertTrue(deleted);

          var found = employeesRepo.selectById(inserted.compositeId(), c);
          assertFalse(found.isPresent());
        });
  }

  @Test
  public void testEmployeesMultipleSameNumber() {
    SqliteTestHelper.run(
        c -> {
          var e1 =
              new EmployeesRow(
                  9999L,
                  "A",
                  "IT",
                  "US-WEST",
                  "Employee A",
                  Optional.empty(),
                  LocalDate.of(2025, 1, 1));
          var e2 =
              new EmployeesRow(
                  9999L,
                  "B",
                  "IT",
                  "US-WEST",
                  "Employee B",
                  Optional.empty(),
                  LocalDate.of(2025, 1, 1));
          var e3 =
              new EmployeesRow(
                  9999L,
                  "C",
                  "IT",
                  "US-WEST",
                  "Employee C",
                  Optional.empty(),
                  LocalDate.of(2025, 1, 1));

          employeesRepo.insert(e1, c);
          employeesRepo.insert(e2, c);
          employeesRepo.insert(e3, c);

          assertTrue(employeesRepo.selectById(new EmployeesId(9999L, "A"), c).isPresent());
          assertTrue(employeesRepo.selectById(new EmployeesId(9999L, "B"), c).isPresent());
          assertTrue(employeesRepo.selectById(new EmployeesId(9999L, "C"), c).isPresent());
        });
  }

  // ==================== Composite FK Relationship ====================

  @Test
  public void testEmployeeDepartmentForeignKey() {
    SqliteTestHelper.run(
        c -> {
          var dept =
              new DepartmentsRow(
                  "FK_TEST", "FK_REGION", "FK Test Dept", Optional.of(new BigDecimal("500000")));
          departmentsRepo.insert(dept, c);

          var emp =
              new EmployeesRow(
                  6001L,
                  "F",
                  "FK_TEST",
                  "FK_REGION",
                  "FK Employee",
                  Optional.of(new BigDecimal("60000")),
                  LocalDate.of(2025, 6, 1));
          var insertedEmp = employeesRepo.insert(emp, c);

          assertEquals(dept.deptCode(), insertedEmp.deptCode());
          assertEquals(dept.deptRegion(), insertedEmp.deptRegion());

          var foundDept =
              departmentsRepo.selectById(
                  new DepartmentsId(insertedEmp.deptCode(), insertedEmp.deptRegion()), c);
          assertTrue(foundDept.isPresent());
        });
  }
}
